### ISSUE TITLE
Let webContents.startDrag() accept a callback that prepares the file

### ISSUE THREAD

**Issue by digitalnature** [electron/electron#9274]
This function currently accepts only file paths. So the file must exist on the local disk. This makes it impossible to build an interface that allows the use of the drag feature to download files from an external source on the desktop. Example- Google Drive.

I think this can easily be implemented by allowing a function to be passed instead. The function handles the download to a temporary location then returns the temporary path to `startDrag()`. And obviously `startDrag()` should only call this function when the item is dropped to avoid unnecessary download.

If it also had support for download progress it would be even better...

**Comment by paulcollett** [comment]
+1 for a callback containing the dropped directory which could be used to save/download a file. Could also be on `endDrag()`?

**Comment by SimulatedGREG** [comment]
It would be fantastic if there was some sort of end callback to know when a file is done transferring. In my case, I'm copying a file to the system temporary directory and using that path with `startDrag()` (I have a specific reason to do this). Once the drag is complete I would like to remove the temporary file, but I don't know when the system is done transferring the file. I could use a `setTimeout`, yet then I would be assuming the file size.

```js
ipcMain.on('ondragstart', (e, path) => {
  e.sender.startDrag({
    file: path,
    icon: join(__static, '/drag.png')
  })

  /* How do I know when the system is done copying the file? */

  removeSync(path)
})
```

**Comment by zcbenz** [comment]
While this would be a very nice feature to have, I'm afraid the ability provided by Chromium does not allow us to implement it. To be able to add this feature, we have to refactor a large part of Chromium's drag-drop code, which simply requires too much resources from us (it might take months).

Closing this as won't fix

**Comment by SimulatedGREG** [comment]
@zcbenz 

Thanks for looking into this. Keep up the awesome work!
